[unitaryhack] Implement single-file HDF5 logger

maxbortone opened the pull request:
This pull request implement a single-file HDF5 logger as described in #1200.

The logger works just as the `JsonLogger`, but flushes logged data and variational state parameters to a single HDF5 file.

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,243 @@
+import os
+import time
+import h5py
```
this would fail if not installed
```suggestion
```

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,243 @@
+from ..stats import Stats
+from ..jax import tree_ravel
+
+
```
hide the import behind a try-catch
```suggestion
try:
    import h5py
except ImportError:
    pass
```

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,243 @@
+            autoflush_cost: Maximum fraction of runtime that can be dedicated to
+                serializing data. Defaults to 0.005 (0.5 per cent)
+        """
+        super().__init__()
```
first thing, try to load h5py. If this fails it will print a nice error.
```suggestion
        import h5py

        super().__init__()
```

PhilipVinc reviewed (commented):
Hdf5py is a too heavy dependency to impose on everyone, and it might be harder to install for the time being on M1 Macs, so I'd prefer we make this an optional dependency.

Please list it under `EXTRA_DEPENDENCIES`.

Then, we need to make sure netket loads fine without `hdf5py` installed (to check, you can just try to uninstall it temporarily and check if `netket` imports fine. 
A trick to do this is to gate the import behind a function. see comments here.

PhilipVinc reviewed (commented):
Hey @maxbortone , thanks for contributing this!

Unfortunately I think the implementation is a bit convoluted and will not work with some of the cases that we support: our loggers should support logging `keys` that are not present in the first iteration, as well as nested dictionaries and keys.

This is particularly relevant for parameters, because we don't want to `tree_ravel` them, as some neural networks have mixes of parameters that can't be ravelled, for example if you mix parameter dtypes.

I think you can take inspiration from `TBLog` (`TensorBoardLogger`), which is in spirit very similar to what you are trying to do here, and should simplify the implementation:

If you just define this function (which is slightly modified from TBLog to work with HDF5 files:

```python
def tree_log(tree, root, data):
    """
    Maps all elements in tree, recursively calling tree_log with a new root string,
    and when it reaches leaves adds them to the `data` inplace.

    Args:
        tree: a pytree where the leaf nodes contain data
        root: the root of the tags used to log to hdf5
        data: an hdf5 file modified in place

    """

    if tree is None:
        return
    
    elif isinstance(tree, list):
        for (i, val) in enumerate(tree):
            tree_log(val, root + f"/{i}", data)
    
    # handle namedtuples
    elif isinstance(tree, list) and hasattr(tree, "_fields"):
        for key in tree._fields:
            tree_log(getattr(tree, key), root + f"/{key}", data)

    elif isinstance(tree, tuple):
        for (i, val) in enumerate(tree):
            tree_log(val, root + f"/{i}", data)

    elif isinstance(tree, dict):
        for key, value in tree.items():
            tree_log(value, root + f"/{key}", data)  # noqa: F722

    elif hasattr(tree, "to_compound"):
        tree_log(tree.to_compound()[1], root, data)  # noqa: F722

    elif hasattr(tree, "to_dict"):
        tree_log(tree.to_dict(), root, data)  # noqa: F722

    else:
        value = np.asarray(tree)
        if root in data:
            f_value = data[root]
            f_value.resize(f_value.shape[0]+1, axis=0)
            f_value[-1] = value
        else:
            maxshape = (None, ) + value.shape
            data.create_dataset(root, data=[value], maxshape = maxshape)
```

if you use it within `__call__` the dictionary to be logged, as well as the parameters of the network, everything should be logged correctly.
For example: `tree_log(log_data, "", self._file_open)`

If you use this approach, you should be able to remove `_init_dataset`.

maxbortone commented:
Thanks for the feedback and suggestion @PhilipVinc. This works and indeed looks much simpler, but requires some modification to add the `iters` dataset to each logged group.

Further, regarding the logging of the variational state. In the desccription of the issue it was outlined that you wish to log the parameters, but with checkpointing in mind I think it might be more helpful to actually log the whole variables dictionary withouth any flattening, so that a model can be restored straight from the data in the log file. What do you think? Should I put this data then under a `variational_state` group, instead of `parameters`?

femtobit commented:
@maxbortone Thanks for working on this!

Regarding how to store the variational parameters:

When outlining the implementation, I indeed did not consider the case where different leaves of the parameter pytree have different data types. It makes sense for us to support properly logging that and storing it as a set of nested HDF5 groups seems to be a sensible solution. There are some corner cases we need to check though. I think we allow arbitrary strings as dictionary keys, so I think this for instance is a valid param pytree in NetKet:
```python3
class M(nn.Module):
    @nn.compact
    def __call__(self, s):
        x = self.param("παράμετρος/1234", nn.initializers.normal(0.1), (1,), complex)
        return x * s
```
```python3
>>> m = M()
>>> m.init(...).unfreeze()
{ "params": { "παράμετρος/1234": ... } }
```
Symbols like slashes or unicode characters could be a problem in HDF5 group names. We could just say we don't support these or we need to do some kind of encoding.

> Should I put this data then under a variational_state group, instead of parameters?

I think for consistency with the `VariationalState` API, the HDF5 group should be called `vairables` if you store the `variables` attribute. (I think we are not 100% consistent with our use of variables vs. parameters in NetKet, but it'd be godd to get it right here.)

PhilipVinc commented:
@maxbortone Ah right. I didn't think about that because in the other case we are using History objects that automatically do that...

In this situation it should be relatively easy to do, you should pass around the recursive `tree_log` an extra keyword argument `iter` and log it as well in the right spots. 

```python
def tree_log(tree, root, data, *, iter=None):
    ...
    elif hasattr(tree, "to_compound"):
        tree_log(tree.to_dict(), f"{root}/iter", data)
        tree_log(tree.to_compound()[1], root, data)  # noqa: F722
    ...
    else:
        value = np.asarray(tree)
        if root in data:
            f_value = data[root]
            f_value.resize(f_value.shape[0]+1, axis=0)
            f_value[-1] = value
        else:
            maxshape = (None, ) + value.shape
            data.create_dataset(root, data=[value], maxshape = maxshape)

        if iter is not None:
            tree_log(iter, f"{root}/iter", data)
```

See [this gist](https://gist.github.com/PhilipVinc/e25cfea5916f86c9a78cf9981441baa9) for the complete function. I haven't tested it, so there might be some typos, but it should work correctly.

@femtobit and @maxbortone about the Parameters: the most compatible way to do this would be to serialise the parameters using flax and store the binary blob in the hdf5 file. We later can provide utility functions to deserialise it.
```python
        binary_data = flax.serialization.to_bytes(variational_state.variables)
        tree_log(binary_data, "variables", file, iter=iter)
```

This is also going to be somewhat annoying for users that want for example to histogram the parameters of the network as a function of time, because that way they will have to deserialise every step and concatenate them. 

For me both solutions would be fine, what do you think @femtobit ?

> EDIT: Okay, it seems unicode is not a problem at least for h5py but slashes result in nested subgroups by default, which could be a problem if {'a/b': ...} get deserialized to {'a': {'b': ...}}. Do we have a way to prevent this?

Escaping slashes as double slashes, I guess? Though I think for now we can have a non-safe implementation and we can leave this as a future TODO. UnitaryHack is ending on Friday.

femtobit commented:
> We later can provide utility functions to deserialise it.

I would prefer it to have the parameters in a format that is easily readable without going through `flax`. This has many benefits. For example if you want to access the evolution of a single parameter (say the first one), in a flattened array that is `parameters[0, :]` while in the serialized case its something like `[flax.serialization.msgpack_restore(x)[0] for x in parameters]`.

I see one sufficiently general way to store the variables in an almost flat fashion while allowing for differing dtypes and weird group names etc.:
1. Flatten (but not ravel!) the pytree, so you get a sequence of arrays, one for each leave. The arrays can have different dtypes.
2. Create one dataset `/variables/{i}` of the appropriate shape and type for each leave where `i` is the leave index.
3. Store the tree structure as an attribute `tree_structure` on `/variables` as a dictionary of the same tree shape as the pytree but with the leaves replaced by their indices. This dictionary can just be encoded as JSON.

For example the pytree
```
{"params": {"Dense/0": array(..., dtype=np.complex128), "Dense/1": array(..., dtype=np.float64)}}
```
is stored as
```
/variables/0    (dtype=complex128)
/variables/1    (dtype=float64)
```
and variables has an attribute
```python3
tree_structure='{"params": {"Dense/0": 0, "Dense/1": 1}}'
```

maxbortone commented:
@PhilipVinc I've refactored it as you suggested, with some modification, as there is a small problem with the code above: it won't work if there are scalar values in the `log_data` dictionary. For example, if we are trying to log the acceptance rate of the sampler, e.g.: 
```
log_data = {'acceptance': np.float64(gs.state.sampler_state.acceptance)}
```
the logger will raise an error because it will try to create a dataset `iter` under the group `acceptance`, which however is a dataset. This is easily solved by changing the structure of `log_data`:
```
log_data = {'acceptance': {'value': np.float64(gs.state.sampler_state.acceptance)}}
```
which will then result in a group `acceptance` with two datasets, `value` and `iter`.

Do you think it should be up to the user to define this structure correctly, or should we handle it properly in the logger? In my previous implementation I was handling this case explicitly, but I'm unsure what is the best way to do it in the recursive function above, since the argument `iter` will also be seen as a scalar.

@femtobit I like your suggestion of splitting `params` and `variables`. I saw that JAX has a `tree_flatten` function that returns a list of flattened leaves and the PyTree structure. Unfortunately, I don't know how to serialize the `PyTreeDef` object, as it can't be cast to a dictionary. Do you know how I can traverse the PyTree and replace the leaves with an incremental index?

However, is there any particular reason why the parameters should be stored as flat arrays? Since HDF5 can handle arrays of different shapes, I don't see the why you wouldn't want to store them in the original shape? Passing the `parameters` dictionary to the `tree_log` function will then log parameters in the appropriate structure and shape.

PhilipVinc commented:
Ah right. I didn't think about this case.
Well it should be very easy to handle it.. just log the value under a `/value` key.  (We want this to be done automatically, not by the user).

You just need to check if there is an `iter` in the `tree_log` function  before logging:
```python
    else:
        if iter is not None:
            tree_log(iter, f"{root}/iter", data)
            root = f"{root}/value"
    
        value = np.asarray(tree)
        if root in data:
            f_value = data[root]
            f_value.resize(f_value.shape[0]+1, axis=0)
            f_value[-1] = value
        else:
            maxshape = (None, ) + value.shape
            data.create_dataset(root, data=[value], maxshape = maxshape)
```

You can also see the [updated gist](https://gist.github.com/PhilipVinc/e25cfea5916f86c9a78cf9981441baa9)

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,179 @@
+    pass
+
+
+def _exists_hdf5(prefix):
+    return os.path.exists(prefix + ".hdf5")
```
Nitpick: Since this function is only used once, can you actually not define it and just paste it's content where you use it?

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,179 @@
+            )
+        mode = _mode_shorthands[mode]
+
+        file_exists = _exists_hdf5(output_prefix)
```
```suggestion
        file_exists = os.path.exists(output_prefix + ".hdf5")
```

maxbortone commented:
@PhilipVinc thanks for the hint, it's been a while since I wrote a recursive function 😅 I've cleaned up the code and wrote some tests, similar to those for the tensorboard logger

PhilipVinc commented on test/logging/test_hdf5_log.py:
```diff
@@ -0,0 +1,54 @@
+    assert len(files) >= 1
+
+
+def test_lazy_init(tmp_path):
+    path = str(tmp_path) + "/dir1"
```
```suggestion
def test_lazy_init(tmp_path):
    # skip test if hdf5py not installed
    pytest.importorskip("hdf5py")

    path = str(tmp_path) + "/dir1"
```

PhilipVinc commented on test/logging/test_hdf5_log.py:
```diff
@@ -0,0 +1,54 @@
+    )
+
+
+def test_hdf5log(vstate, tmp_path):
+    path = str(tmp_path) + "/dir1/dir2"
```
those tests require `hdf5py` installed, but we don't run all tests with all extra dependencies, so we need to skip it in this case..

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,173 @@
+
+    def __call__(self, step, log_data, variational_state):
+        if self._writer is None:
+            self._writer = h5py.File(self._file_name, self._file_mode)
+        tree_log(log_data, "data", self._writer, iter=step)
```
just a newline to make this a bit more readable...
```suggestion
            self._writer = h5py.File(self._file_name, self._file_mode)

        tree_log(log_data, "data", self._writer, iter=step)
```

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,173 @@
+        if self._writer is None:
+            self._writer = h5py.File(self._file_name, self._file_mode)
+        tree_log(log_data, "data", self._writer, iter=step)
+        if self._save_params and step % self._save_params_every == 0:
```
`step` is an integer when running `VMC` but it's a floating point (the time) when running `TDVP`, so this code won't work.

The way to do this that we used in to keep an internal counter of the steps, `self._steps_notsaved_params` and increase it by 1 at the end of the `__call__` function. See for example [StateLog](https://github.com/netket/netket/blob/6e0d9eef36fe60bc923a399195cfca08f7f31766/netket/logging/state_log.py#L155)

```suggestion
        if self._steps_notsaved_params and step % self._save_params_every == 0:
```

PhilipVinc commented on test/logging/test_hdf5_log.py:
```diff
@@ -0,0 +1,54 @@
+    log.flush()
+    del log
+
+    files = glob.glob(path + "/*")
+    assert len(files) >= 1
```
can you also try opening the file and assert that there `energy` and `complex` in the hdf5py file and they have length 30?

PhilipVinc reviewed (commented):
Ok... we're almost there, the last few fixes and we should be good.

The only thing that I'm not convinced about is the params/variables serialization, but that's just a design issue and I'm discussing it offline with @femtobit . I'm unsure of what would be the best approach...

femtobit commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,173 @@
+        output_prefix: str,
+        mode: str = "write",
+        save_params: bool = True,
+        save_params_every: int = 50
```
I think `save_params_every=1` is indeed a sensible default. Saving all the states can help a lot with debugging (or computing observables after the optimization etc.) and the only reason not to do it that I can see is if it costs to much disk space, at which point users can easily adjust this option.

maxbortone commented:
I've fixed the imports in the tests and added checks on the length of the logged data, added the internal counter, changed the default setting and fixed the formatting with black.

Let me know what you decide regarding the logging of the variational state, I still have time to work on this today.

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,174 @@
+            variables = variational_state.variables.unfreeze()
+            params = variables.pop("params")
+            binary_data = to_bytes(variables)
+            tree = {"variables": binary_data, "parameters": params}
```
variables is the union of parameters and everything else, in NetKet we use the term `model_state` for the everything else so it would be better to call it like that.
```suggestion
            tree = {"model_state": binary_data, "parameters": params}
```

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,174 @@
+            params = variables.pop("params")
+            binary_data = to_bytes(variables)
+            tree = {"variables": binary_data, "parameters": params}
+            tree_log(tree, "variational_state", self._writer, iter=step)
```
as we're logging everything together, I fear the `tree_log` above would store multiple times the `iter` entry so let's just store it once manually.
```suggestion
            tree_log(tree, "variational_state", self._writer)
            tree_log(step, "variational_state/iter", self._writer)
```

PhilipVinc commented on test/logging/test_hdf5_log.py:
```diff
@@ -0,0 +1,68 @@
+    assert len(files) >= 1
+
+    f = h5py.File(files[0], "r")
+    energy = np.array(f["data/Energy/value"])
+    complex = np.array(f["data/complex/value"])
+    params = np.array(f["variational_state/parameters/Dense/kernel/value"])
+    assert energy.shape[0] == 30
+    assert complex.shape[0] == 30
+    assert params.shape[0] == 30
```
adapt to the changes above
```suggestion
    energy = np.array(f["data/Energy/value"])
    complex = np.array(f["data/complex/value"])
    params = np.array(f["variational_state/parameters/Dense/kernel"])
    iters = np.array(f["variational_state/iter"])
    assert energy.shape[0] == 30
    assert complex.shape[0] == 30
    assert params.shape[0] == 30
    assert iters.shape[0] == 30
```

PhilipVinc commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,174 @@
+                "`[x]`(fail)."
+            )
+        mode = _mode_shorthands[mode]
+
```
I should add this to the other loggers as well, it always gets me. So since we're here... let's make it work also in the case that one specifies the `.hdf5` as well.
```suggestion

    if output_prefix.endswith(".hdf5"):
        output_prefix = output_prefix[:-5]

```

PhilipVinc reviewed (commented):
@maxbortone Thanks a lot for your contribution! I think this is done now! I added the tag to signal that you get the bounty.

Regarding the serialization of model_state/variables, I think what you implemented is actually a good compromise. Let's see how it plays out in the wild.

I also left a bunch of minor fixes that I think you can just merge from the GitHub interface that should polish this a bit more.

Before actually merging, the last two things to do would be to add this logger to the documentation by adding it's name to [`docs/api/logging.md`](https://github.com/netket/netket/blob/master/docs/api/logging.md) under the others and add a line to the [`CHANGELOG`](https://github.com/netket/netket/blob/master/CHANGELOG.md) mentioning that this logger has been added under `new features`.

maxbortone commented:
Great! Thanks a lot for the quick review process, it was a pleasure helping out. I'll add the logger to the documentation and update the changelog.

maxbortone commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,184 @@
+            data.create_dataset(root, data=[value], maxshape=maxshape)
+
+
+class HDF5Log(RuntimeLog):
```
Should I remove the inheritance on the runtime logger?

PhilipVinc replied:
while you're at it can you also run black on the branch? I think some of my commits broke the formatting.

maxbortone replied:
Yup, will do

femtobit commented on netket/logging/hdf5_log.py:
```diff
@@ -0,0 +1,184 @@
+
+        self._prefix = output_prefix
+        self._file_mode = mode
+        self._file_name = self._prefix + ".hdf5"
```
I think `.h5` is the more common suffix for HDF5 files than `.hdf5`. Since both are fine though, maybe we should just let users decide what to use and not add an extension? (The logger only creates one file anyways, it is okay if the user defined prefix already contains the file extension.)

In that case, I'd just remove the `self._prefix` field and only store `self._file_name`

maxbortone replied:
It might be useful to have a standard file extension for checkpointing at some point later. Maybe we can jsut stick with the most common one and ditch `.hdf5` in favor of `.h5`?

femtobit reviewed (changes requested):
@PhilipVinc Maybe we should merge this under `netket.experimental` for now just in case we find out we want to handle some aspect (in particular the serialization of `.variables`) differently in the near future? (This would of course still qualify for the bounty.)

PhilipVinc commented:
About the experimental... Fine by me. If nobody complains in 2 months we can move it to the main netket

femtobit commented:
It seems currently it is not possible to read from the HDF5 file while the corresponding driver is still running.

Reopening the file every time by creating a temporary `h5py.File` object in "append" mode within `__call__` helps. Is there a more elegant way to make sure the file is readable by other processes between logging calls? I would have hoped calling `.flush()` after each step would do that, but I quickly tested it and that doesn't suffice.

PhilipVinc commented:
@femtobit not easy but possible https://docs.h5py.org/en/stable/swmr.html

maxbortone commented:
I've moved the logger to the experimental submodule and added basic SWMR support. Now the output file should be readable during optimization. However, no new dataset or groups can be added, as pointed out by @femtobit. Maybe SWMR should better be an optional setting for now?

femtobit commented:
> Maybe SWMR should better be an optional setting for now?

We could add an option to turn it off. However, safety of the log data in case of a crash should be on by default in the `HDF5Log`, which is currently only provided when SWMR is turned on.

@maxbortone Is it possible to add the workaround of just closing the file, temporarily reopening it without SWMR, add the groups, and then re-enable SWMR? Since groups/datasets are not created often, this should be okay, even though it is a bit ugly.

PhilipVinc commented on netket/experimental/logging/hdf5_log.py:
```diff
@@ -0,0 +1,186 @@
+
+    def __call__(self, step, log_data, variational_state):
+        if self._writer is None:
+            self._writer = h5py.File(self._file_name, self._file_mode, libver='latest')
+            self._writer.swmr_mode = True
```
```suggestion
            self._writer = h5py.File(self._file_name, self._file_mode)
```

PhilipVinc commented:
@femtobit , @maxbortone The PR is good for me, and I would like to merge this. In particular because that way we can revert additional changes if needed.

Discussions about how to support `swmr` mode (which will be some hack/workaround probably reopening the file) and other stuff can be done in a new issue/PR?

maxbortone commented:
@PhilipVinc sounds good to me 👍🏼

maxbortone commented:
Thanks to you guys, glad I could help out a bit. Happy to contribute more in the near future!

nathanshammah commented:
@maxbortone it seems like you have not signed up on unitaryHACK, please sign up to be able to claim the bounty https://airtable.com/shrJeycewBFqdot2B

maxbortone commented:
@nathanshammah Thanks for pointing that out. I think I signed up with a different email/account at the beginning
